localstack/services/awslambda/invocation/version_manager.py

```diff
@@ -191,39 +213,119 @@ def stop(self) -> None:
             state=State.Inactive, code=StateReasonCode.Idle, reason="Shutting down"
         )
         self.shutdown_event.set()
-        self.destination_execution_pool.shutdown(
-            cancel_futures=True
-        )  # TODO: give it a grace period waiting, otherwise fail
+        self.provisioning_pool.shutdown(wait=False, cancel_futures=True)
+        self.destination_execution_pool.shutdown(wait=False, cancel_futures=True)
 
         self.queued_invocations.put(QUEUE_SHUTDOWN)
         self.available_environments.put(QUEUE_SHUTDOWN)
+
+        futures_exenv_shutdown = []
+        for environment in list(self.all_environments.values()):
+            futures_exenv_shutdown.append(
+                self.execution_env_pool.submit(self.stop_environment, environment)
+            )
         if self.invocation_thread:
             try:
                 self.invocation_thread.join(timeout=5.0)
                 LOG.debug("Thread stopped '%s'", self.function_arn)
             except TimeoutError:
                 LOG.warning("Thread did not stop after 5s '%s'", self.function_arn)
-        for environment in list(self.all_environments.values()):
-            self.stop_environment(environment)
+
+        concurrent.futures.wait(futures_exenv_shutdown, timeout=3)
+        self.execution_env_pool.shutdown(wait=False, cancel_futures=True)
         self.log_handler.stop()
         get_runtime_executor().cleanup_version(self.function_version)
 
     def update_provisioned_concurrency_config(self, provisioned_concurrent_executions: int) -> None:
-        self.provisioned_concurrent_executions = provisioned_concurrent_executions
-        # TODO initialize/destroy runners if applicable
+        """
+        TODO: implement update while in progress (see test_provisioned_concurrency test)
+        TODO: loop until diff == 0 and retry to remove/add diff environments
+        TODO: alias routing & allocated
+        TODO: ProvisionedConcurrencyStatusEnum.FAILED
+        TODO: status reason
+
+        :param provisioned_concurrent_executions: set to 0 to stop all provisioned environments
+        """
+
+        if self.provisioned_state == ProvisionedConcurrencyStatusEnum.IN_PROGRESS:
```

Comment:
Doesn't make too much sense comparing the container with the state here but actually comparing the state will break it as well, as IN_PROGRESS is the default state when creating the provisioned_state in provider.py.

We probably should check if it is not None, and set the container afterwards, instead of doing it in provider.py.